### DISCUSSION TITLE
Heads-up: the build system can now build in parallel

### DISCUSSION THREAD

**Discussion by McDutchie** [ksh93/ksh#774] [New features]
See ced499a3ce39dd2aa404e7ef85855ee0718a9ec1 (and follow-up commits) :-)

Now on the dev and 1.0 branches (but not in the 1.0.10 release that is just out -- I need this tested first).

Please test this as described in the linked commit and let me know your thoughts and experiences.

Ping @hyenias, @JohnoKing, @posguy99, anyone else interested

**Comment by posguy99** [comment]
I have to say, it's **really** freaky watching `package` run things in parallel. ^_^

This M2 Max has 12 cores (8 performance and 4 efficiency) and 32 gb RAM.  But it doesn't gain much past 4-5 parallel jobs.  It'd be interesting to know what the macOS kernel is actually doing, whether the parallelism is being scheduled on the performance cores or the efficiency cores.


```
time (bin/package clean && bin/package make)
# src/Mamfile: 12-17: make announce_build_done
# KornShell Version AJM 93u+m/1.1.0-alpha+ac8fe7fd 2024-08-01
# Type 'bin/package use' to try it out
package: make done at Fri Aug  2 21:14:24 PDT 2024 in /Users/mwilson/src/ksh/arch/darwin.arm64-64

real	1m45.250s
user	1m16.373s
sys	0m36.054s


time (bin/package clean && bin/package make -j5)
Total errors: 0
# src/Mamfile: 12-17: make announce_build_done
# KornShell Version AJM 93u+m/1.1.0-alpha+ac8fe7fd 2024-08-01
# Type 'bin/package use' to try it out
package: make done at Fri Aug  2 21:11:30 PDT 2024 in /Users/mwilson/src/ksh/arch/darwin.arm64-64

real	0m54.688s
user	1m18.419s
sys	0m42.340s


time (bin/package clean && bin/package make -j12)
# src/Mamfile: 12-17: make announce_build_done
# KornShell Version AJM 93u+m/1.1.0-alpha+ac8fe7fd 2024-08-01
# Type 'bin/package use' to try it out
package: make done at Fri Aug  2 21:09:55 PDT 2024 in /Users/mwilson/src/ksh/arch/darwin.arm64-64

real	0m52.317s
user	1m21.851s
sys	0m46.099s
```

**Reply by McDutchie** [reply]
> I have to say, it's **really** freaky watching `package` run things in parallel. ^_^

Yeah, I enjoy that too ;-)

I now have a solution running that makes log output usable again when running in parallel. Each shell action redirects its output to a temp file and mamake dumps it all at once when the job finishes. Since mamake itself never runs in parallel (only the shell actions it spawns do that), this creates usable log output while running in parallel. Rule output appears out of order (depending on when they finish) but the dependency blocking ensures it's still a sensible order. Commit coming up soon.

> This M2 Max has 12 cores (8 performance and 4 efficiency) and 32 gb RAM. But it doesn't gain much past 4-5 parallel jobs.

Yeah, that's because many things in this code base depend on each other -- often on iffe feature tests that take a long time to complete; the math.sh test in ksh93 is the worst one for that. So there's only so much potential for parallelisation. When a rule depends on an action that hasn't finished yet, it has to wait for it, there's simply no way around that. To have better parallelisation, we'd have to split those large jobs into multiple smaller jobs that can run in parallel. But that is often *very* complicated.

**Comment by JohnoKing** [comment]
It's working perfectly fine on my system (Arch Linux, Ryzen 5600 w/ HT and PBO).

GCC 14.1.1:
```
# No parallelism
package: make done at Fri Aug  2 21:45:07 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	1m26.823514s
cpu	1m20.600670s

# -j12
package: make done at Fri Aug  2 21:43:29 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	0m37.414166s
cpu	1m44.754415s

# -j6
package: make done at Fri Aug  2 21:48:45 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	0m38.251227s
cpu	1m28.116085s
```

tcc 0.9.28rc mob@e02eec6b:
```
# No parallelism
package: make done at Fri Aug  2 21:49:56 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	0m31.277583s
cpu	0m26.585680s

# -j12
package: make done at Fri Aug  2 21:50:39 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	0m14.875750s
cpu	0m29.765943s

# -j6
package: make done at Fri Aug  2 21:51:15 PDT 2024 in /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64

real	0m15.128604s
cpu	0m27.580403s
```

FWIW I also tried switching out `fork` for `posix_spawn` in mamake to see if it would make any difference, but there was no difference in performance (maybe it would help on a slower system). Patch below:

<details>
<summary>mamake-posix_spawn.patch</summary>

```patch
diff --git a/src/cmd/INIT/mamake.c b/src/cmd/INIT/mamake.c
index 0a577c969..ac9df61ca 100644
--- a/src/cmd/INIT/mamake.c
+++ b/src/cmd/INIT/mamake.c
@@ -156,6 +156,10 @@ static const char usage[] =
 #include <sys/wait.h>
 #include <time.h>
 
+#if _POSIX_VERSION >= 200112L
+#include <spawn.h>
+#endif
+
 #if !_PACKAGE_ast
 #include <errno.h>
 #include <signal.h>
@@ -1368,9 +1372,28 @@ static int execute(Rule_t *r, char *s)
 {
 	int	stat;
 	pid_t	pid;
-
-	if (!state.shell && (!(state.shell = getval(state.vars, "SHELL")) || !strcmp(state.shell, sh)))
-		state.shell = sh;
+#if _POSIX_VERSION >= 200112L
+	int spawn_err;
+	posix_spawnattr_t attr;
+	char *script_argv[] = { state.shell, "-c", s, NULL };  /* set at declaration time to avoid malloc */
+
+	/* posix_spawn is supported starting from POSIX.1-2001 */
+	spawn_err = posix_spawnattr_init(&attr);
+	if(spawn_err)
+	{
+		errno = spawn_err;
+		report(3, strerror(errno), "posix_spawnattr_init() failed", 0);
+	}
+	report(-5, s, "exec", 0);
+	spawn_err = posix_spawn(&pid, state.shell, NULL, &attr, script_argv, environ);
+	posix_spawnattr_destroy(&attr);
+	if(spawn_err)
+	{
+		errno = spawn_err;
+		report(3, strerror(errno), "posix_spawn() failed", 0);
+	}
+#else
+	/* fallback for ancient systems without posix_spawn(3) */
 	pid = fork();
 	if (pid < 0)
 		report(3, strerror(errno), "fork() failed", 0);
@@ -1382,6 +1405,7 @@ static int execute(Rule_t *r, char *s)
 			exit(127);
 		exit(126);
 	}
+#endif
 	/* parent */
 	/* run job in background if wanted & possible (never for -r) */
 	if (r && state.maxjobs > 1 && state.strict >= 5 && !state.recurse)
@@ -2866,6 +2890,13 @@ int main(int argc, char **argv)
 	auto_updprev = search(state.vars, "?", 1);
 	auto_making->value = auto_prev->value = auto_allprev->value = auto_updprev->value = empty;
 
+	/*
+	 * select the shell
+	 */
+
+	if (!state.shell && (!(state.shell = getval(state.vars, "SHELL")) || !strcmp(state.shell, sh)))
+		state.shell = sh;
+
 	/*
 	 * grab the command line targets and variable definitions
 	 */

```

</details>

**Edit:** After testing my old laptop using an AMD V140 (single core, no parallelization), the difference in compilation speed averaged about 1 second faster using `posix_spawn` (06m12s vs 06m11s averaged over four compiles for each method), which is statistically insignificant.

**Reply by McDutchie** [reply]
Thanks for testing!

> FWIW I also tried switching out `fork` for `posix_spawn` in mamake to see if it would make any difference, but there was no difference in performance (maybe it would help on a slower system).

I made mamake count all its fork(2) calls and rebuilt from scratch:
```
$ grep 'forks==' arch/darwin.arm64-64/lib/package/gen/make.out                                                                     
[DEBUG] forks==48
[DEBUG] forks==724
[DEBUG] forks==8
[DEBUG] forks==19
[DEBUG] forks==62
[DEBUG] forks==5
[DEBUG] forks==103
[DEBUG] forks==24
[DEBUG] forks==2
```
That's 995 in total.

The following times 995 forks of ksh on my Apple M1 system (a shell background process must always fork as it contains a copy of the parent process):

```
$ time (set +m; typeset -i i; for ((i=0; i<995; i++)); do : & wait; done)

real	0m00.54s
user	0m00.06s
sys	0m00.42s
```

So yeah, it's hardly going to make the difference. :-)

Also, it looks like a significant of a fork(2) call runs in parallel, i.e., control is passed back to the caller well before the fork is finished and the child starts. If we don't wait for `: &` to finish:

```
$ time (set +m; typeset -i i; for ((i=0; i<995; i++)); do : & done)

real	0m00.20s
user	0m00.06s
sys	0m00.51s
```

So, the time cost of fork(2) is that much smaller when running in parallel.

**Comment by hyenias** [comment]
I primarily have low end single board computers and entry level computers that I work with and none of them have hyper threading. All of them have at least 4 CPUs with much smaller amounts of CPU L1, L2, and L3 caches. The new -j parallelism works without issues on all systems tested. However, after running the numbers, I would recommend that the formula be equal to half of available CPUs on entry level CPUs as using more provides little improvement on non-desktop/laptop class CPUs. Attempting to use more CPUs in parallel than hardware supports was slower. I believe the max used CPU cores should be one less than MAX allowing a CPU for operating system, sshd, and etc.

| Unit | System|  CPU | -j# | real | user | sys |
|---|----------|---|---|----|-------|-------|
| 1 | i386-64 | 4 | NA |  2m49.674s | 2m12.680s | 0m30.866s |
| 1 | i386-64 | 4 | -j2 |  1m34.872s | 2m25.094s | 0m32.434s |
| 1 | i386-64 | 4 | -j3 |  1m27.405s | 2m31.361s | 0m33.766s |
| 1 | i386-64 | 4 | -j4 |  1m32.535s | 2m37.879s | 0m34.401s |
| 1 | i386-64 | 4 | -j5 |  1m36.214s | 2m40.789s | 0m35.164s |
| 2 | linux.arm64-64| 4 | NA | 12m56.464s | 10m49.297s | 2m3.050s |
| 2 | linux.arm64-64| 4 | -j2 | 7m18.875s | 12m12.716s | 2m17.037s |
| 2 | linux.arm64-64| 4 | -j3 | 7m20.165s | 13m13.564s | 2m45.739s |
| 2 | linux.arm64-64| 4 | -j4 | 7m2.568s | 13m6.143s | 2m20.993s |
| 2 | linux.arm64-64| 4 | -j5 | 7m10.792s| 13m13.951s | 2m19.691s |
| 3 | linux.arm64-64| 4 | NA | 13m4.162s | 9m59.455s | 2m46.906s |
| 3 | linux.arm64-64| 4 | -j2 | 7m16.156s | 11m16.270s | 3m1.606s |
| 3 | linux.arm64-64| 4 | -j3 | 6m53.534s | 12m1.304s | 3m11.058s |
| 3 | linux.arm64-64| 4 | -j4 | 7m0.264s | 12m11.904s | 3m10.328s |
| 3 | linux.arm64-64| 4 | -j5 | 7m8.443s | 12m16.835s | 3m9.331s |
| 4 | linux.arm| 8 | NA | 7m26.044s | 5m22.653s | 1m55.211s |
| 4 | linux.arm| 8 | -j2 | 4m8.641s | 6m0.594s | 2m11.081s |
| 4 | linux.arm| 8 | -j4 | 3m48.865s | 6m41.777s | 2m26.962s |
| 4 | linux.arm| 8 | -j7 | 3m44.000s| 7m49.209s | 2m37.700s |

Example of testing on a unit:
```
$ time bin/package quiet make
package: initialize the /tmp/ksh/arch/linux.arm64-64 view
package: update /tmp/ksh/arch/linux.arm64-64/bin/mamprobe
package: update /tmp/ksh/arch/linux.arm64-64/lib/probe/C/make/probe
package: update /tmp/ksh/arch/linux.arm64-64/bin/mamake
package: make output captured in /tmp/ksh/arch/linux.arm64-64/lib/package/gen/make.out

real    13m4.162s
user    9m59.455s
sys     2m46.906s

$ time bin/package quiet make -j2
package: initialize the /tmp/ksh/arch/linux.arm64-64 view
package: update /tmp/ksh/arch/linux.arm64-64/bin/mamprobe
package: update /tmp/ksh/arch/linux.arm64-64/lib/probe/C/make/probe
package: update /tmp/ksh/arch/linux.arm64-64/bin/mamake
package: make output captured in /tmp/ksh/arch/linux.arm64-64/lib/package/gen/make.out

real    7m16.156s
user    11m16.270s
sys     3m1.606s
```

**Comment by McDutchie** [comment]
New commit: b541c9b7ef930b6909d741de14d4eb232cea79db

mamake now saves up log output by default and dumps it when each job is finished. Pass `-c` if you'd prefer to watch the chaos instead :)

**Comment by ormaaj** [comment]
I haven't rebuilt ksh since the dynamic lib push went in because I assume this changes things with new libs and maybe headers, like for writing plugins with shlib maybe? I will try building manually and see if I can figure it out from the result.

I normally use an ebuild I adapted from gentoo's but don't think there are any gentoo devs that follow this repo closely unless something breaks.
